cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataEntity.java

```diff
@@ -170,11 +173,21 @@ public Builder appendAsType(String key, String value) {
      * the CDAP entity
      */
     public MetadataEntity build() {
-      if (parts.isEmpty() || type == null || type.isEmpty()) {
-        throw new IllegalArgumentException("key-value pair and type must be specified");
+      if (parts.isEmpty()) {
+        throw new IllegalArgumentException("key-value pair must be specified");
+      }
+      if (type == null) {
+        // traverse till the last key and make that the type
+        parts.keySet().forEach(x -> type = x);
       }
       validateHierarchy();
-      return new MetadataEntity(Collections.unmodifiableMap(parts), type);
+      return new MetadataEntity(new LinkedHashMap<>(parts), type);
+    }
+
+    private void validateKey(String key) {
+      if (key == null || key.isEmpty()) {
```

Comment:
should every key also be unique? Seems that is required

Reply:
Done.